---
title: "Discovering regulatory switches in complex biological networks"
author: "Ram√≥n Reszat, Luca Farinola, Patrick Mertens, Alaric Deboilly"
date: "2023-01-11"
output: html_document
---

```{r DGEA, include=FALSE}
library(ggplot2) # create plots for the presentation
library(pcaMethods) # Bioconductor PCA methods
library(limma) # differential gene expression analysis (DGEA)
library(biomaRt) # download gene annotations from Ensembl
```

## Load MAGnet and Ensembl Datasets
First, load the original [MAGnet dataset](https://www.med.upenn.edu/magnet) and information on the study participants.
```{r MAGnet, warning=FALSE}
setwd('Data')

# transcriptome of each sample from the NCBI MAGnet dataset
gxData <- read.delim('MAGNET_GeneExpressionData_CPM_19112020.txt', header=TRUE, sep='\t', row.names=1)

# participant information for the tissue samples in the dataset
sampleInfo <- read.csv('MAGNET_SampleData_18112022.csv',header=TRUE, stringsAsFactor=TRUE)

# exon length of each gene for normalization of CPM to TPM
exonLength <- read.delim('MAGNET_exonLengths.txt', header=TRUE, sep='\t', row.names=1)

# load the list of genes that are suspected biomarkers for type 2 diabetes
biomarkers <- read.delim('MAGNET_Biomarkers_11012023.txt', header=FALSE, sep='\t')
```

Additionally, download gene annotations from the [Ensembl](http://www.ensembl.org/index.html) database using the biomaRt API.
```{r}
# connect to the Ensembl genes database and select homo sapiens as a species
ensembl = useEnsembl(biomart="genes", dataset="hsapiens_gene_ensembl", mirror='useast')

# download the Ensembl IDs for the genes selected from literature research
biomarkers <- getBM(attributes=c('ensembl_gene_id', 'hgnc_symbol'), filters='hgnc_symbol', values=biomarkers, mart=ensembl)
```

## Differential Gene Expression Analysis

```{r design}
# if the baseline level is the healthy group limma will create the design matrix correctly
sampleInfo$etiology <- factor(sampleInfo$etiology, levels = c("NF", "DCM", "HCM", "PPCM"))

# creates dummy matrix and correct for confounding factors
design <- model.matrix(~0+etiology + gender + age, data=sampleInfo)
```

```{r contrasts}
# create contrasts for the linear model from the design matrix
contrasts <- makeContrasts(contrasts=c("etiologyDCM-etiologyNF"), levels=colnames(design))
```

$$
\text{Gene Expression} = 
\begin{cases}
\beta_0 + \beta_1 & \text{if case is DCM} \\
\beta_0 & \text{if case is NF}
\end{cases} + 
\begin{cases}
\beta_2 + \beta_3 & \text{if person is male} \\
\beta_2 & \text{if person is female} 
\end{cases} + 
\beta_4 \text{ Age}
$$

```{r model}
# fit linear regression for each gene
model <- lmFit(gxData, design)

# compute coefficients for each case-control pair
model  <- contrasts.fit(model, contrasts)

# Benjamini-Hochberg correction (FDR)
model <- eBayes(model, trend=TRUE)

head(model$coefficients)
```

```{r genes}
# generate top-tables of p-values and log-fold changes
genes <- topTable(model, coef=1, n=Inf, adjust="fdr")

genes <- genes[genes$adj.P.Val < 5E-2,]
genes <- genes[abs(genes$logFC) > 1.5,]
head(genes)
```

## Gene Correlation Network Analysis

```{r, include=FALSE}
library(Hmisc) # Harrell miscellaneous: rcorr function
library(igraph) # graph and network analysis
library(NbClust)
library(factoextra)
```

```{r, include=FALSE}
# load all R functions from the SwitchMiner in the project directory
switchminer <- list.files(path=getwd(), pattern="*.R$", recursive=TRUE)
sapply(switchminer, source)
```

```{r}
# select genes from differential gene expression analysis and samples according to the study design
data <- gxData[rownames(genes), (sampleInfo$etiology == 'NF') | (sampleInfo$etiology == 'DCM')]

# options: c("pearson", "spearman"), c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none")
edge_list <- computeCorrelation(data, "spearman", "fdr")

threshold_prc_corr <- 0.9

threshold_corr <- quantile(edge_list$correlation, threshold_prc_corr)
threshold_corr <- round(as.numeric(threshold_corr),digits=4)
```

```{r}
corr <- ggplot(edge_list, aes(x=correlation)) +
  geom_histogram() +
  labs(
    title="",
    subtitle="",
    tag="Figure 1",
    x = "",
    y = ""
  )

corr + theme_bw()
```

```{r}
min_rho <- 0.1
max_rho <- 0.9
step_rho <- 0.05
threshold_pval_adj <- 0.05

# TODO: network integrity plot
network_integrity <- checkNetworkIntegrity(edge_list, min_rho, max_rho, step_rho, threshold_pval_adj)
```

```{r}
network <- buildCorrelationNetwork(edge_list, threshold_corr, threshold_pval_adj)
w_adj <- createWeightedAdjMatrix(network)

getScreePlot(w_adj)
```

```{r}
# TODO: wrap into nice output object
clustering <- getClustering(w_adj, 4, 100, 5)

num_node <- nrow(w_adj)
num_edge <- nrow(network)

res <- list(network = network,
            num_node = num_node, 
            num_edge = num_edge,
            w_adj = w_adj,
            threshold_corr = threshold_corr,
            idx = clustering$idx,
            cluster_size = clustering$size,
            WSS = clustering$WSS, 
            TWSS = clustering$TWSS)
```

## Mining Regulatory Switches

```{r, warning=FALSE}
APCC <- computeAPCC(w_adj)
res_deg <- computeDegree(w_adj,res$idx)
hub_class <- getHubClassification(APCC,res_deg$deg)

cartography <- buildCartography(genes,res$idx,res_deg,APCC,hub_class,res)

getAPCCdistribution(cartography$attribute)
```

```{r}
attribute_switch <- getSwitch(cartography$attribute)
getHeatCartography(cartography$attribute)
```